README
NAME
    Data::Checker - a framework for checking data validity

SYNOPSIS
       use Data::Checker;

       $obj = new Data::Checker;

DESCRIPTION
    A commonly performed task is to have a set of data that you want to
    validate. Given a set of elements, you want to test each to make sure
    that it is valid, and then break the set into two groups: the group that
    is valid, and the group that is not. With the group that is not valid,
    you usually want an error message associated with that element so you
    can see why it is not valid.

    Although this is an extremely common task, there isn't a convenient
    framework for expressing these tests in, which means that every time you
    want to do this kind of testing, you not only have to write the
    functions for doing the tests, you also have to write the entire
    framework as well.

    This module was written to provide the framework around the tests. A
    number of common test functions are provided, or you can write your own,
    and the framework will take care of the rest.

    The framework includes the following commonly desired functionality:

    Automatic handling of the testing
        A list of elements is passed in to the framework, and it will
        automatically apply the tests and split the elements into sets of
        passing and failing elements.

    Running tests in parallel
        Many times, testing a piece of data may take a significant amount of
        time, and running them in parallel can greatly speed up the process.

        This framework allows you to run any number of the tests in
        parallel, or you can run them serially one at a time.

    Support for warnings and information messages
        Sometimes you want some tests to produce warnings or just
        informational messages for an element, but to still consider them as
        having passed the test.

        The level for each test can be specified so that a failure produces
        an informational notice, a warning, or an error. Only an error means
        that the element fails the test.

BASE METHODS
    new
           $obj = new Data::Checker;

        This creates a new data check framework.

    version
           $vers = $obj->version;

        This returns the version of this module.

    parallel
           $obj->parallel($n);

        In many cases, tests can be run in parallel to speed things up. By
        default, all tests will be run serially (one at a time) but that
        behavior can be changed using this method. $n must be a positive
        integer (or zero):

           $n=1  All tests are run serially. This is the default.

           $n>1  $n tests will run at a time.  If there are more
                 elements than this, one will have to finish before
                 another will start.

           $n=0  All of the elements will be tested at the same time.

    check
           ($pass,$fail,$warn,$info) = $obj->check($data,$check,$check_opts);

        This is the function which actually performs the checks. It takes a
        set of elements ($data) and verifies them using the checks specified
        by $check and $check_opts. It returns a list of elements that pass
        the check and a list that fail the check. In addition, informational
        notes and warnings about the elements may also be returned.

        The data is passed in as a single data structure ($data) as
        described below in the SPECIFYING DATA section.

        $check specifies what function to use to perform a check. It will be
        used to test an individual element to determine whether it passes or
        fails a check. This is described below in the CHECK FUNCTION
        section.

        $check_opts is a hashref that contains options specifying exactly
        how the check is to be performed, and it will be passed to the check
        function. This is described more fully below in the CHECK OPTIONS
        section.

SPECIFYING DATA
    Data is passed in to the check method in one of two forms.

    The simplest form is a listref. For example:

       $data = [ 'cow', 'horse', 'zebra', 'oak' ]

    Many tests do not require any more than this. For these, elements that
    pass are returned also as a listref. Order is NOT preserved in the
    output ($pass and $fail).

    Some tests however rely on a description of each element, and for these,
    the data is passed in as a hashref where each key is one data element
    and the value is a description of the elements (which will typically be
    a hashref, but might be a scalar, a listref, or some other type of
    description, and will be documented with the function doing the check).

    For example:

       $data = { 'apple'  => { 'type'  => 'fruit',
                               'color' => 'red' },
                 'pear'   => { 'type'  => 'fruit',
                               'color' => 'yellow' },
                 'bean'   => { 'type'  => 'vegetable',
                               'color' => 'green' }
               }

    As mentioned, the exact form of the description will be documented with
    the function that is used to do the checks.

    When data is passed in as a hashref, the list of elements that pass is
    also a hashref with the description fully preserved.

CHECK FUNCTION
    All checks are performed by a function which takes a single element and
    tests it to see if it passes. It may perform only a single check on an
    element, or multiple checks.

    All check functions take the same set of arguments, and all return the
    same set of values.

    The $check argument in the check method provides a pointer to where the
    check function can be found.

    $check can be a coderef, in which case you are passing the check
    function in directly. Alternately, $check can be a string naming the
    check function, or the namespace where it is found.

    If $check is a string, the Data::Checker framework will look for a check
    function based on that string. As an example, if $check is Foo::Bar, the
    following locations will be examined to see if they are a function:

       Foo::Bar
       Foo::Bar::check
       CALLER::Foo::Bar
       CALLER::Foo::Bar::check
       Data::Checker::Foo::Bar
       Data::Checker::Foo::Bar::check

    where CALLER is the package of the calling routine. The first one which
    refers to a function will be used. The appropriate module will be loaded
    as necessary.

    A check function always called as follows:

       ($element,$err,$warn,$info) =
         FUNCTION($obj,$element,$description,$check_opts);

    The arguments to the check function are:

    $obj
        $obj is the Data::Checker object that was created, and is passed in
        to provide the check function some useful methods provided by the
        framework. These functions are described below in the CHECK FUNCTION
        METHODS

    $element, $description
        $element is the element being tested, and $description is the
        description of that element.

        If the list of elements was specified as a listref, $element will be
        one value from that listref and $description will be undef.

        If the list of elements was specified as a hashref, $element will be
        one of the keys from that hashref and $description will be the value
        of that key.

    $check_opts
        $check_opts is the hashref that was passed in to the check method
        and is described in the next section.

    The check function always returns the following values:

    $element
        This is the element that was passed in as an argument. It must be
        returned so that when parallel testing is done, the parent can
        easily determine which element was being checked by a finished
        thread.

    $err
        This is a listref of error messages. If this is undefined or empty,
        then the element passed the test.

    $info, $warn
        These are listrefs of informational messages and warnings about this
        element. These are optional.

CHECK OPTIONS
    Options may be passed in to the check function as a hashref. The form of
    the hashref (what keys/values are allowed) is documented with the check
    function, but the general form is:

       $check_opts = { GLOBAL_OPT_1 => GLOBAL_VAL_1,
                       GLOBAL_OPT_2 => GLOBAL_VAL_2, ...

                       CHECK_A      => { OPT_A1 => VAL_A1,
                                         OPT_A2 => VAL_A2, ... }
                       CHECK_B      => { OPT_A1 => VAL_A1,
                                         OPT_A2 => VAL_A2, ... }
                       ... }

    There are two types of keys in $check_opts: ones which sets global
    options (which apply to all possible checks that could be done), and
    ones which define exactly what types of checks are performed and options
    that apply only to that check.

    All check specific options will override a global option.

    The following options are standard:

    level => err, warn, info
        The level option (which defaults to 'err') can be set to 'warn' or
        'info'. If it is, then any element which fails this check will
        produce the appropriate type of message. It will only result in a
        failure if it is set to 'err'.

    negate => 1
        The negate option can be set to negate the test (i.e. what would
        have been deemed a success it actually a failure and vice versa.

    message => [ STRING, STRING, ... ]
        The message to return if a check fails. The string __ELEMENT__ will
        be replaced by the element being checked.

    For example, doing DNS checks, you might want to specify exactly what
    server to use, and you might want to check that a host is defined in DNS
    (and produce an error if it is not), and warn if it does not have a
    unique IP. This might be done by passing in:

       $check_opts = { 'nameservers'  => 'my.dns.server',
                       'dns'          => undef,
                       'unique'       => { 'level' => 'warn' } }

CHECK FUNCTION METHODS
    In addition to the base methods listed above, the Data::Checker object
    also includes the following methods which are intended to be called
    inside a check function.

    check_performed
           $flag = $obj->check_performed($check_opts,$label);

        This checks $check_opts for the existance of a key named $label
        indicating that that check should be performed.

    check_level
           $level = $obj->check_level($check_opts [,$label]);

        Check to see what level ('err', 'info', or 'warn');

    check_option
           $val = $obj->check_option($check_opts,$opt [,$default [,$label]]);

        This returns the value of the given option ($opt) for this check
        ($label).

        If the option is not found, $default is returned (if it is given).

    check_message
           $obj->check_message($check_opts,$label,$element,$default_message,
                               $level,$err,$warn,$info);

        This produces a message indicating that the check failed and stores
        it in the appropriate listref.

        If the 'message' option is available, that message is used.
        Otherwise, $default_message will be used.

        The message can be a string or a listref (a multi-line message). The
        string __ELEMENT__ will be replaced by the element being tested.

    check_value
           $obj->check_value($check_opts,$label,$element,$value,
                             $std_fail,$negate_fail,
                             $err,$warn,$info);

        This will test to see if a check passed or failed. It takes a value
        ($value) and if it evaluates to true, then by default the check
        passes (unless the 'negate' option is present in which case it
        fails).

        The $std_fail is a message (either a string or a listref of strings)
        that will be given when the check fails. $negate_fail is a similar
        message that will be given when the check fails but the 'negate'
        option is present.

        $err, $warn, and $info are listrefs containing the messages.

        If $err is non-empty, an error has occurred.

        If the $negate_fail empty is empty, the 'negate' option will be
        ignored. This is typically used to test an element to see if it is
        the right type of data for this check. If it isn't, other types of
        checks are typically not able to run.

        If $label is empty, the test is always performed.

KNOWN BUGS AND LIMITATIONS
    None known.

SEE ALSO
    Data::Checker::DNS
        Predefined DNS checks.

    Data::Checker::Ping
        Predefined checks to see if a host reponds to pings.

    Data::Checker::IP
        Predefined checks to see if an element is a valid IP.

LICENSE
    This script is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.

AUTHOR
    Sullivan Beck (sbeck@cpan.org)



Changes
Written by:
   Sullivan Beck (sbeck@cpan.org)
Any suggestions, bug reports, or donations :-) should be sent to me.

1.09

1.08  2016-11-22
  - Cleaned up some tests.  RT 118881

1.07  2016-11-02
  - Disabled PING tests during installation (they are too unstable)
    RT 118541
  - Added date checks.
  - Fixed parallel tests.

1.06  2015-04-16
  - Improved POD formatting.

1.05  2014-11-24
  - Fixed dns.t file.  Thanks to two CPAN testers (Nigel Horne and
    Chris Williams) for assisting me in finding the offending test.

1.04  2014-09-15
  - Fixed a problem related to the new .git directory that prevented
    it from installing correctly.

1.03  2014-09-11
  - Minor changes to some tests
  - Put distro on GitHub.  Gabor Szabo

1.02  2014-09-09
  - The DNS test had 2 problems.  The 1.01 release only fixed one
    of them.  This adds a workaround for some broken versions of
    Net::DNS.  (RT 98598)

1.01  2014-09-08
  - Add examples to docs
  - Fixed a DNS test (RT 98598)
  - Added Data::Checker::IP
  - Modified Data::Checker::DNS to support IPs
  - Modified check_value method to support non-negatable tests

1.00  2014-01-30
  - Initial release.



Build.PL
######################################################################
# Build.PL for Data::Checker
######################################################################

use strict;
use warnings;
use 5.008;
use Module::Build 0.21;

my %module_prereq = (
   'perl'                     => '5.008',
   'Module::Loaded'           => '0',
   'Net::DNS'                 => '0',
   'Net::Ping'                => '0',
   'Net::Ping::External'      => '0',
   'NetAddr::IP'              => '4.073',
   'NetAddr::IP::Lite'        => '0',
   'Parallel::ForkManager'    => '0',
);

my %script_prereq = (
);

my %test_prereq = (
   'Test::Inter'              => '0',
   'Test::More'               => '0',
);

my %config_prereq = (
   'Module::Build'            => '0.21',
);

my %build_prereq = %config_prereq;
if ($Module::Build::VERSION < 0.4004) {
   %build_prereq = (%config_prereq,%test_prereq);
}

my $class = Module::Build->subclass(
   class  => "Module::Build::Custom",
   code   => q { sub ACTION_html { my $self = shift ; $self->SUPER::ACTION_html ( @_ ) if ( $ENV{ MAKE_HTML } ) ; } }
);

my $build = $class->new(
      module_name        => "Data::Checker",
      dist_version       => "1.06",
      dist_abstract      => "a framework for checking data validity",
      license            => "perl",
      dist_author        => "Sullivan Beck <sbeck\@cpan.org>",

      requires           => \%module_prereq,
      ($Module::Build::VERSION >= 0.4004
       ? (test_requires  => \%test_prereq)
       : ()),
      ($Module::Build::VERSION >= 0.30
       ? (configure_requires  => \%config_prereq)
       : ()),
      build_requires     => \%build_prereq,
      recommends         => \%script_prereq,
      sign               => 1,

      ($Module::Build::VERSION >= 0.28
       ? (meta_merge => {
             resources => {
                repository => "https://github.com/SBECK-github/Data-Checker",
             }
          })
       : ()),
);

$build->create_build_script;


Makefile.PL
######################################################################
# Makefile.PL for Data::Checker
######################################################################

use strict;
use warnings;
use 5.008;
use ExtUtils::MakeMaker 6.30;

my %IsWin32 = map { $_ => 1 } qw( MSWin32 NetWare symbian );
my $IsWin32 = 1   if ($IsWin32{ $^O });
my $Pvers   = $];
my $ConvVers= $ExtUtils::MakeMaker::VERSION;

my %module_prereq = (
   'Date::Manip'              => '0',
   'Module::Loaded'           => '0',
   'Net::DNS'                 => '0',
   'Net::Ping'                => '0',
   'Net::Ping::External'      => '0',
   'NetAddr::IP'              => '4.073',
   'NetAddr::IP::Lite'        => '0',
   'Parallel::ForkManager'    => '0',
);

my %script_prereq = (
);

my %test_prereq = (
   'Test::Inter'              => '0',
   'Test::More'               => '0',
);

my %config_prereq = (
   'ExtUtils::MakeMaker'      => '6.30',
);

%module_prereq = (%module_prereq,%script_prereq);
my $EU_MM_V    = $ExtUtils::MakeMaker::VERSION;
$EU_MM_V       =~ s/_//g;
if ($EU_MM_V < 6.64) {
   %config_prereq = (%config_prereq,%test_prereq);
}

WriteMakefile(
   NAME             => "Data::Checker",
   VERSION          => "1.09",
   ABSTRACT         => "a framework for checking data validity",
   LICENSE          => "perl",
   ($EU_MM_V >= 6.48
    ? (MIN_PERL_VERSION => "5.008")
    : ()),
   AUTHOR           => "Sullivan Beck (sbeck\@cpan.org)",
   "dist"           => {COMPRESS=>"gzip",SUFFIX=>"gz"},
   PL_FILES         => {},
   PREREQ_PM        => \%module_prereq,
   ($EU_MM_V >= 6.52
    ? (CONFIGURE_REQUIRES => \%config_prereq)
    : ()),
   ($EU_MM_V >= 6.64
    ? (TEST_REQUIRES   => \%test_prereq)
    : ()),
   ($EU_MM_V > 6.45
    ? (META_ADD => {
          provides     => {
             'Data::Checker' => {
                file    => 'Checker.pm',
                version => '1.09',
             },
             'Data::Checker::DNS' => {
                file    => 'DNS.pm',
                version => '1.09',
             },
             'Data::Checker::Date' => {
                file    => 'Date.pm',
                version => '1.09',
             },
             'Data::Checker::IP' => {
                file    => 'IP.pm',
                version => '1.09',
             },
             'Data::Checker::Ping' => {
                file    => 'Ping.pm',
                version => '1.09',
             },

          },
       }
      )
    : ()),

   ($EU_MM_V > 6.46
    ? (META_MERGE => {
          'meta-spec'  => {
             version   => 2,
          },
          resources    => {
             repository => {
                type => 'git',
                web  => 'https://github.com/SBECK-github/Data-Checker',
                url  => 'git://github.com/SBECK-github/Data-Checker.git',
             },
          },
       }
      )
    : ()),




);


lib/Data/Checker/IP.pm
package Data::Checker::IP;
# Copyright (c) 2014-2016 Sullivan Beck. All rights reserved.
# This program is free software; you can redistribute it and/or modify it
# under the same terms as Perl itself.

###############################################################################

require 5.008;
use warnings 'all';
use strict;
use NetAddr::IP::Lite qw(:nofqdn);
use NetAddr::IP 4.073;

our($VERSION);
$VERSION='1.09';

###############################################################################
###############################################################################

sub check {
   my($obj,$element,$desc,$check_opts) = @_;
   my $err    = [];
   my $warn   = [];
   my $info   = [];

   #
   # Must be a valid IP,
   #
   # Must be any of the forms supported by NetAddr::IP.
   #

   my $ip_obj = new NetAddr::IP $element;
   my $valid  = ($ip_obj ? 1 : 0);
   $obj->check_value($check_opts,undef,$element,$valid,
                     "Not a valid IP",undef,
                     $err,$warn,$info);
   return ($element,$err,$warn,$info)  if (@$err);

   #
   # Check IP version.
   #

   my $vers   = $ip_obj->version();
   if ($obj->check_performed($check_opts,'ipv4')) {
      my $is4  = ($vers == 4);
      $obj->check_value($check_opts,'ipv4',$element,$is4,
                        "IPv4 IP required",
                        "Non-IPv4 IP required",
                        $err,$warn,$info);
      return ($element,$err,$warn,$info)  if (@$err);
   }

   if ($obj->check_performed($check_opts,'ipv6')) {
      my $is6  = ($vers == 6);
      $obj->check_value($check_opts,'ipv6',$element,$is6,
                        "IPv6 IP required",
                        "Non-IPv6 IP required",
                        $err,$warn,$info);
      return ($element,$err,$warn,$info)  if (@$err);
   }

   #
   # in_network
   #

   my $net_obj;
   if ($obj->check_performed($check_opts,'in_network')) {
      my $val   = $obj->check_option($check_opts,'network',undef,'in_network');
      $net_obj  = new NetAddr::IP $val;

      # Must be a valid IP
      my $valid  = ($net_obj ? 1 : 0);
      $obj->check_value($check_opts,undef,$element,$valid,
                        "in_network must be a valid IP",undef,
                        $err,$warn,$info);
      return ($element,$err,$warn,$info)  if (@$err);

      # Must be the same version
      my $v     = $net_obj->version();
      if ($v != $vers) {
         $obj->check_value($check_opts,undef,$element,0,
                           "in_network and IP must both be IPv4 or IPv6",undef,
                           $err,$warn,$info);
         return ($element,$err,$warn,$info);
      }

      # Must contain network info
      my $mask  = $net_obj->masklen();
      if ( ($vers == 4  &&  $mask == 32) ||
           ($vers == 6  &&  $mask == 128) ) {
         $obj->check_value($check_opts,undef,$element,0,
                           "in_network must be a valid network IP",undef,
                           $err,$warn,$info);
         return ($element,$err,$warn,$info);
      }

      my $flag  = $net_obj->contains($ip_obj);
      $obj->check_value($check_opts,'in_network',$element,$flag,
                        "IP not in network",
                        "IP contained in network",
                        $err,$warn,$info);

      return ($element,$err,$warn,$info)  if (@$err);
   }

   #
   # network_ip, broadcast_ip
   #

   my $chk_net_ip   = ($obj->check_performed($check_opts,'network_ip') ? 1 : 0);
   my $chk_broad_ip = ($obj->check_performed($check_opts,'broadcast_ip') ? 1 : 0);
   if ( ($chk_net_ip  ||  $chk_broad_ip)  &&  ! $net_obj) {
      my $mask = $ip_obj->masklen();

      if ( ($vers == 4  &&  $mask == 32) ||
           ($vers == 6  &&  $mask == 128) ) {
         $obj->check_value($check_opts,undef,$element,0,
                           "IP must include network information for " .
                           "network/broadcast check",undef,
                           $err,$warn,$info);
         return ($element,$err,$warn,$info);
      }

      $net_obj = $ip_obj;
   }

   if ($chk_net_ip) {
      my $net_ip  = $net_obj->network->addr();
      my $ip      = $ip_obj->addr();
      $valid      = ($ip eq $net_ip);
      $obj->check_value($check_opts,'network_ip',$element,$valid,
                        "Network IP required",
                        "Non-network IP required",
                        $err,$warn,$info);
      return ($element,$err,$warn,$info)  if (@$err);
   }

   if ($chk_broad_ip) {
      my $broad_ip  = $net_obj->broadcast->addr();
      my $ip        = $ip_obj->addr();
      $valid        = ($ip eq $broad_ip);
      $obj->check_value($check_opts,'broadcast_ip',$element,$valid,
                        "Broadcast IP required",
                        "Non-broadcast IP required",
                        $err,$warn,$info);
      return ($element,$err,$warn,$info)  if (@$err);
   }

   return ($element,$err,$warn,$info);
}

1;
# Local Variables:
# mode: cperl
# indent-tabs-mode: nil
# cperl-indent-level: 3
# cperl-continued-statement-offset: 2
# cperl-continued-brace-offset: 0
# cperl-brace-offset: 0
# cperl-brace-imaginary-offset: 0
# cperl-label-offset: 0
# End:


lib/Data/Checker/Date.pm
package Data::Checker::Date;
# Copyright (c) 2014-2016 Sullivan Beck. All rights reserved.
# This program is free software; you can redistribute it and/or modify it
# under the same terms as Perl itself.

###############################################################################

require 5.008;
use warnings 'all';
use strict;
use Date::Manip;

our($VERSION);
$VERSION='1.09';

###############################################################################
###############################################################################

sub check {
   my($obj,$element,$desc,$check_opts) = @_;
   my $err    = [];
   my $warn   = [];
   my $info   = [];

   #
   # Must be a valid date.
   #
   # Must be any of the forms supported by Date::Manip.
   #

   my $date   = ParseDate($element);
   my $valid  = ($date ? 1 : 0);
   $obj->check_value($check_opts,undef,$element,$valid,
                     "Not a valid date",undef,
                     $err,$warn,$info);
   return ($element,$err,$warn,$info);
}

1;
# Local Variables:
# mode: cperl
# indent-tabs-mode: nil
# cperl-indent-level: 3
# cperl-continued-statement-offset: 2
# cperl-continued-brace-offset: 0
# cperl-brace-offset: 0
# cperl-brace-imaginary-offset: 0
# cperl-label-offset: 0
# End:


lib/Data/Checker/Ping.pm
package Data::Checker::Ping;
# Copyright (c) 2013-2016 Sullivan Beck. All rights reserved.
# This program is free software; you can redistribute it and/or modify it
# under the same terms as Perl itself.

###############################################################################

require 5.008;
use warnings 'all';
use strict;
use Net::Ping;
use Net::Ping::External;

our($VERSION);
$VERSION='1.09';

###############################################################################
###############################################################################

sub check {
   my($obj,$element,$desc,$check_opts) = @_;
   my $err  = [];
   my $warn = [];
   my $info = [];

   if (! defined $check_opts) {
      $check_opts = { 'tcp' => undef };
   }

   # Do the pings

   foreach my $proto (qw(tcp udp icmp stream syn external)) {
      next  if (! $obj->check_performed($check_opts,$proto));

      my $ping     = Net::Ping->new($proto);
      my $timeout  = $obj->check_option($check_opts,'timeout',5,$proto);
      my $up       = ($ping->ping($element,$timeout) ? 1 : 0);

      $obj->check_value($check_opts,$proto,$element,$up,
                        "Host does not respond to $proto pings",
                        "Host does respond to $proto pings",
                        $err,$warn,$info);
      return ($element,$err,$warn,$info)  if (@$err);
   }

   return ($element,$err,$warn,$info);
}

1;
# Local Variables:
# mode: cperl
# indent-tabs-mode: nil
# cperl-indent-level: 3
# cperl-continued-statement-offset: 2
# cperl-continued-brace-offset: 0
# cperl-brace-offset: 0
# cperl-brace-imaginary-offset: 0
# cperl-label-offset: 0
# End:


lib/Data/Checker.pm
package Data::Checker;
# Copyright (c) 2013-2016 Sullivan Beck. All rights reserved.
# This program is free software; you can redistribute it and/or modify it
# under the same terms as Perl itself.

###############################################################################

require 5.008;
use warnings 'all';
use strict;
use Module::Loaded;
use Parallel::ForkManager 0.7.6;

our($VERSION);
$VERSION='1.09';

###############################################################################
# BASE METHODS
###############################################################################

sub version {
   my($self) = @_;

   return $VERSION;
}

sub new {
   my($class,@args) = @_;

   my $self = {
               'parallel'     => 1,
              };

   bless $self, $class;

   return $self;
}

# Some checks can be run in parallel.  For these, passing in $n
# has the following effect:
#   $n = 0  : all of them will run simultaneously
#   $n = 1  : only one check at a time
#   $n > 1  : $n checks at a time
#
sub parallel {
   my($self,$n) = @_;

   if (defined($n)  &&  $n =~ /^\d+$/) {
      $n += 0;
   } else {
      warn "WARNING: Invalid argument to Data::Checker::parallel\n";
      return;
   }

   $$self{'parallel'} = $n + 0;
}

###############################################################################

sub check {
   my($self,$data,$type,$opts) = @_;

   # Check for data

   my (%data,$wantlist);
   if      (ref($data) eq 'ARRAY') {
      %data     = (map { $_,undef } @$data);
      $wantlist = 1;
   } elsif (ref($data) eq 'HASH') {
      %data     = %$data;
      $wantlist = 0;
   } else {
      die "ERROR: invalid data passed to Data::Checker::check\n";
   }

   # Find the check function

   my $func;
   if (! defined($type)) {
      die "ERROR: invalid check function passed to Data::Checker::check\n";

   } elsif (ref($type) eq 'CODE') {
      $func = $type;

   } else {
      my $caller = ( caller() )[0];

      TRY:
      foreach my $name ("${type}",
                        "${type}::check",
                        "${caller}::${type}",
                        "${caller}::${type}::check",
                        "Data::Checker::${type}",
                        "Data::Checker::${type}::check",
                       ) {

         # Ignore the case where $name does not have '::' because that means
         # we called it with the name of a function in the CALLER namespace
         # (so it'll get handled by one of the "${caller}::" cases, or $type
         # is a sub-namespace of Data::Checker.

         next  if ($name !~ /^(.*)::(.+)$/);
         my($mod) = ($1);
         $mod = "main"  if (! defined $mod);

         # Try loading the module (but not main:: or CALLER::

         if ($mod ne 'main'  &&
             $mod ne $caller  &&
             ! is_loaded($mod)) {
            next TRY  if (! eval "require $mod");
         }

         # Look for the function

         no strict 'refs';
         if (defined &{$name}) {
            $func = \&{$name};
            last TRY;
         }
      }

      die "ERROR: no valid check function passed to Data::Checker::check\n"
        if (! defined $func);
   }

   # Call parallel or serial check

   if ($$self{'parallel'} != 1) {
      return $self->_check_parallel(\%data,$wantlist,$func,$opts);
   } else {
      return $self->_check_serial(\%data,$wantlist,$func,$opts);
   }
}

sub _check_parallel {
   my($self,$data,$wantlist,$func,$opts) = @_;
   my(%pass,%fail,%info,%warn);
   my @ele      = keys %$data;
   my $max_proc = ($$self{'parallel'} > 1 ? $$self{'parallel'} : @ele);

   my $manager = Parallel::ForkManager->new($max_proc);
   $manager->run_on_finish
     (
      sub {
         my($pid,$exit_code,$id,$signal,$core_dump,$funcdata) = @_;
         my($ele,$err,$warn,$info) = @$funcdata;

         if (defined($err)  &&  @$err) {
            $fail{$ele} = $err;
         } else {
            $pass{$ele} = $$data{$ele};
         }

         if (defined($warn)  &&  @$warn) {
            $warn{$ele} = $warn;
         }
         if (defined($info)  &&  @$info) {
            $info{$ele} = $info;
         }
      });

   ELE:
   foreach my $ele (sort keys %$data) {
      $manager->start and next;

      my($element,$err,$warn,$info) = &$func($self,$ele,$$data{$ele},$opts);

      $manager->finish(0,[$element,$err,$warn,$info]);
   }

   $manager->wait_all_children();

   if ($wantlist) {
      my @pass = sort keys %pass;
      return (\@pass,\%fail,\%warn,\%info);
   } else {
      return (\%pass,\%fail,\%warn,\%info);
   }
}

sub _check_serial {
   my($self,$data,$wantlist,$func,$opts) = @_;
   my(%pass,%fail,%info,%warn);

   ELE:
   foreach my $ele (sort keys %{ $data }) {
      my($element,$err,$warn,$info) = &$func($self,$ele,$$data{$ele},$opts);

      if (defined($err)  &&  @$err) {
         $fail{$ele} = $err;
      } else {
         $pass{$ele} = $$data{$ele};
      }

      if (defined($warn)  &&  @$warn) {
         $warn{$ele} = $warn;
      }
      if (defined($info)  &&  @$info) {
         $info{$ele} = $info;
      }
   }

   if ($wantlist) {
      my @pass = sort keys %pass;
      return (\@pass,\%fail,\%warn,\%info);
   } else {
      return (\%pass,\%fail,\%warn,\%info);
   }
}

###############################################################################
# CHECK OPTIONS METHODS
###############################################################################

sub check_performed {
   my($self,$check_opts,$label) = @_;

   return 1  if (exists $$check_opts{$label});
   return 0;
}

sub check_option {
   my($self,$check_opts,$opt,$default,$label) = @_;

   if (defined $label  &&
       exists $$check_opts{$label}  &&
       exists $$check_opts{$label}{$opt}) {
      return $$check_opts{$label}{$opt};

   } elsif (exists $$check_opts{$opt}) {
      return $$check_opts{$opt};

   } else {
      return $default;
   }
}

sub check_level {
   my($self,$check_opts,$label) = @_;
   return $self->check_option($check_opts,'level','err',$label);
}

sub check_message {
   my($self,$check_opts,$label,$element,$message,$level,$err,$warn,$info) = @_;

   my $mess = $self->check_option($check_opts,'message',$message,$label);
   my @mess;
   if (ref($mess) eq 'ARRAY') {
      @mess = @$mess;
   } else {
      @mess = ($mess);
   }
   foreach my $m (@mess) {
      $m =~ s/__ELEMENT__/$element/g;
   }

   if ($level eq 'info') {
      push(@$info,@mess);
   } elsif ($level eq 'warn') {
      push(@$warn,@mess);
   } else {
      push(@$err,@mess);
   }
}

sub check_value {
   my($self,$check_opts,$label,$element,$value,$std_fail,$negate_fail,
      $err,$warn,$info) = @_;

   while (1) {

      # We perform the check if the $label check is performed, or if
      # there is no label.

      my $do_check = 1  if (! $label  ||
                            $self->check_performed($check_opts,$label));
      last  if (! $do_check);

      # Find the severity level and negate options (negate will never
      # occur if we didn't pass in a negate_fail message).

      my $level  = $self->check_level($check_opts,$label);
      my $negate = $self->check_option($check_opts,'negate',0,$label);
      $negate    = 0  if (! defined($negate_fail));

      # Check the value.

      if (! $negate  &&  ! $value) {
         $self->check_message($check_opts,$label,$element,$std_fail,
                              $level,$err,$warn,$info);
      } elsif ($negate  &&  $value) {
         $self->check_message($check_opts,$label,$element,$negate_fail,
                              $level,$err,$warn,$info);
      }

      last;
   }

   return ($element,$err,$warn,$info);
}

1;
# Local Variables:
# mode: cperl
# indent-tabs-mode: nil
# cperl-indent-level: 3
# cperl-continued-statement-offset: 2
# cperl-continued-brace-offset: 0
# cperl-brace-offset: 0
# cperl-brace-imaginary-offset: 0
# cperl-label-offset: 0
# End:
